File: src/main/java/de/kevcodez/metronom/parser/AlertParser.java
package de.kevcodez.metronom.parser;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import de.kevcodez.metronom.converter.AlertConverter;
import de.kevcodez.metronom.model.alert.Alert;
import de.kevcodez.metronom.utility.Exceptions;
import de.kevcodez.metronom.utility.WebsiteSourceDownloader;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static de.kevcodez.metronom.utility.WebsiteSourceDownloader.getCookiesFromUrl;

/**
 * Class to parse alert notifications from the Metronom SOAP endpoint.
 *
 * @author Kevin Grüneberg
 */
@Component
public class AlertParser {

    public static final String METRONOM_ALERT_URL = "https://www.der-metronom.de/livedata/etc?type=troublelist&_=1538252265246";

    private static ObjectMapper objectMapper = new ObjectMapper();

    @Autowired
    private AlertConverter alertConverter;

    @Autowired
    private WebsiteSourceDownloader websiteSourceDownloader;

    /**
     * Parses the alerts from the Metronom SOAP endpoint.
     *
     * @return list of alerts
     */
    public List<Alert> parseAlerts() {
        Map<String, String> cookies = getCookiesFromUrl("https://www.der-metronom.de/ueber-metronom/wer-wir-sind");
        String pageSource = websiteSourceDownloader.getSource(METRONOM_ALERT_URL, cookies);
        return parseAlertsFromSource(pageSource);
    }

    private List<Alert> parseAlertsFromSource(String pageSource) {
        try {
            JsonNode mainJsonNode = objectMapper.readTree(pageSource);
            JsonNode alertJsonNode = mainJsonNode.get("stoerung");

            List<Alert> alerts = new ArrayList<>();
            alertJsonNode.forEach(jsonAlert -> {
                Alert alert = alertConverter.convert(jsonAlert);

                alerts.add(alert);
            });

            return alerts;
        } catch (IOException exc) {
            throw Exceptions.unchecked(exc);
        }
    }

}

File: src/main/kotlin/de/kevcodez/metronom/parser/AlertParser.kt
package de.kevcodez.metronom.parser

import com.fasterxml.jackson.databind.ObjectMapper
import de.kevcodez.metronom.converter.AlertConverter
import de.kevcodez.metronom.model.alert.Alert
import de.kevcodez.metronom.utility.WebsiteSourceDownloader
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.*

/**
 * Class to parse alert notifications from the Metronom SOAP endpoint.
 *
 * @author Kevin Grüneberg
 */
@Component
class AlertParser @Autowired constructor(
    private val alertConverter: AlertConverter,
    private val websiteSourceDownloader: WebsiteSourceDownloader,
    private val objectMapper: ObjectMapper
) {

    /**
     * Parses the alerts from the Metronom SOAP endpoint.
     *
     * @return list of alerts
     */
    fun parseAlerts(): List<Alert> {
        val cookies = websiteSourceDownloader.getCookiesFromUrl("https://www.der-metronom.de/ueber-metronom/wer-wir-sind")
        val pageSource = websiteSourceDownloader.getSource(METRONOM_ALERT_URL, cookies)
        return parseAlertsFromSource(pageSource)
    }

    private fun parseAlertsFromSource(pageSource: String): List<Alert> {
        val mainJsonNode = objectMapper.readTree(pageSource)
        val alertJsonNode = mainJsonNode.get("stoerung")

        val alerts = ArrayList<Alert>()
        alertJsonNode.forEach { jsonAlert ->
            val alert = alertConverter.convert(jsonAlert)

            alerts.add(alert)
        }

        return alerts
    }

    companion object {

        const val METRONOM_ALERT_URL = "https://www.der-metronom.de/livedata/etc?type=troublelist&_=1538252265246"

    }

}
